### PR TITLE
MVS extension

### PR DESCRIPTION
**Pull request by midlik**
# Description
MolViewSpec (MVS) extension allows loading MVS view descriptions into Mol*. The new functionality is available:
- in the viewer via URL parameters `mvs-url`, `mvs-data`, `mvs-format`,
- in the viewer via drag&drop for `.mvsj` files,
- programmatically via `MVSData` (including data builder and validator) and `loadMVS`.

## Actions

- [x] Added description of changes to the `[Unreleased]` section of `CHANGELOG.md`
- [x] Updated headers of modified files
- [x] Added my name to `package.json`'s `contributors`

### CODE REVIEW

File: src/extensions/mvs/additions/annotation-prop.ts [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+    constructor(
+        public data: AnnotationData,
+        public schema: AnnotationSchema,
+    ) { }
+
+    /** Create a new `Annotation` based on specification `spec`. Use `file` if provided, otherwise download the file.
+     * Throw error if download fails or problem with data. */
+    static async fromSpec(ctx: CustomProperty.Context, spec: AnnotationSpec, file?: AnnotationFile): Promise<Annotation> {
+        file ??= await getFileFromSource(ctx, annotationSourceFromSpec(spec));
```
Comment (dsehnal) [inline src/extensions/mvs/additions/annotation-prop.ts:150]: Cool, didn't know this operator existed.

File: src/extensions/mvs/camera.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        const cell = plugin.state.data.cells.get(structureNodeSelector.ref);
+        structure = cell?.obj?.data;
+        if (!structure) console.warn('Focus: no structure');
+        if (!(structure instanceof Structure)) {
+            console.warn('Focus: cannot apply to a non-structure node');
+            structure = undefined;
+        }
+    }
+    const boundingSphere = structure ? Loci.getBoundingSphere(Structure.Loci(structure)) : getPluginBoundingSphere(plugin);
+    if (boundingSphere && plugin.canvas3d) {
```
Comment (dsehnal) [inline src/extensions/mvs/camera.ts:58]: Would `this.plugin.managers.camera.focusSphere()` (or some other variant) work here?

And if not, possible extend that method with optional arguments to behave the same?

File: src/extensions/mvs/helpers/indexing.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        const atomsById = new NumberMap<number, ElementIndex>(nAtoms + 1);
+        const atomsByIndex = new NumberMap<number, ElementIndex>(nAtoms);
+
+        for (let iChain = 0 as ChainIndex; iChain < nChains; iChain++) {
+            chainsByLabelEntityId.add(label_entity_id.value(iChain), iChain);
+            chainsByLabelAsymId.add(label_asym_id.value(iChain), iChain);
+            chainsByAuthAsymId.add(auth_asym_id.value(iChain), iChain);
+
+            const iResFrom = h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain]];
+            const iResTo = h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1]] ?? nResidues;
```
Comment (dsehnal) [inline src/extensions/mvs/helpers/indexing.ts:56]: I think this is a bug. If you want exclusive "end residue" this should be 

```ts
// atomEnd = h.chainAtomSegments.offsets[iChain + 1] is the 1st atom of the "next chain"
// lastResidue = h.residueAtomSegments.index[atomEnd - 1]
// residueUpperBound = lastResidue + 1
//   =>
h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1] - 1] + 1
```

The joys of indexing :)

File: src/extensions/mvs/helpers/param-definition.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+function parseMaybeInt(input: string): number | undefined {
+    if (input.trim() === '-') return -0;
+    const num = parseInt(input);
+    return isNaN(num) ? undefined : num;
+}
+function stringifyMaybeInt(num: number | undefined): string {
+    if (num === undefined) return '';
+    if (Object.is(num, -0)) return '-';
+    return num.toString();
+}
```
Comment (dsehnal) [inline src/extensions/mvs/helpers/param-definition.ts:24]: Interesting hack. I think a better solution solution could be to add `ParamDefinition.OptionalNumber`  and bake this logic into the React control itself. Though for now, this is fine too.

File: src/extensions/mvs/helpers/param-definition.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+/** Similar to `PD.Text` but leaving empty field in UI is treated as `undefined` */
+export function PD_MaybeString(defaultValue?: string, info?: PD.Info): PD.Base<string | undefined> {
```
Comment (dsehnal) [inline src/extensions/mvs/helpers/param-definition.ts:27]:
```suggestion
export function MaybeStringParamDefinition(defaultValue?: string, info?: PD.Info): PD.Base<string | undefined> {
```

File: src/extensions/mvs/helpers/param-definition.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
+ *
+ * @author Adam Midlik <midlik@gmail.com>
+ */
+
+import { ParamDefinition as PD } from '../../../mol-util/param-definition';
+
+
+/** Similar to `PD.Numeric` but allows leaving empty field in UI (treated as `undefined`) */
+export function PD_MaybeInteger(defaultValue?: number, info?: PD.Info): PD.Base<number | undefined> {
```
Comment (dsehnal) [inline src/extensions/mvs/helpers/param-definition.ts:11]:
```suggestion
export function MaybeIntegerParamDefinition(defaultValue?: number, info?: PD.Info): PD.Base<number | undefined> {
```

File: src/extensions/mvs/helpers/utils.ts [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+ * Equivalent to `dst.push(...src)`, but avoids storing element on call stack. Faster that `extend` from Underscore.js.
+ * `extend(a, a)` will double the array
+ */
+export function extend<T>(dst: T[], src: ArrayLike<T>): void {
+    const offset = dst.length;
+    const nCopy = src.length;
+    dst.length += nCopy;
+    for (let i = 0; i < nCopy; i++) {
+        dst[offset + i] = src[i];
+    }
```
Comment (dsehnal) [inline src/extensions/mvs/helpers/utils.ts:121]: Just use `.push` here.

File: src/extensions/mvs/helpers/utils.ts [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        result[i] = start + i;
+    }
+    return result;
+}
+
+/** Copy all elements from `src` to the end of `dst`.
+ * Equivalent to `dst.push(...src)`, but avoids storing element on call stack. Faster that `extend` from Underscore.js.
+ * `extend(a, a)` will double the array
+ */
+export function extend<T>(dst: T[], src: ArrayLike<T>): void {
```
Comment (dsehnal) [inline src/extensions/mvs/helpers/utils.ts:115]:
```suggestion
export function extendArray<T>(dst: T[], src: ArrayLike<T>): void {
```

File: src/extensions/mvs/helpers/utils.ts
Comment (dsehnal) [inline src/extensions/mvs/helpers/utils.ts:1]: I think it would be better if you move these functions to appropriate files in `mol-util` (e.g. `array.ts`, `object.ts`, etc). Or possibly create a new file if the function doesn't fit.

File: src/extensions/mvs/additions/annotation-color-theme.ts
Comment (dsehnal) [inline src/extensions/mvs/additions/annotation-color-theme.ts:1]: Can we get a better name for  the folder than `additions`? Maybe `components`?

**Review by dsehnal — COMMENTED** [review]
Good job! It's quite hard to review this in detail, but when I spotted something that seemed out of place I commented on that. We might do an extra PR or two with bugfixes after this to iron bugs.

**Comment by arose** [comment]
Nice, please give me a week to review this, thanks.

File: src/extensions/mvs/helpers/indexing.ts (hunk 3 above)
Comment (midlik) [reply src/extensions/mvs/helpers/indexing.ts:56]: I'm taking "the first residue of the next chain", you're taking "the last residue of the current chain + 1".
Won't they always be the same?
```
h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1]] ?? nResidues
h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1] - 1] + 1
```

File: src/apps/viewer/app.ts [hunk 9]
```diff
@@ -467,6 +470,30 @@ export class Viewer {
         return { model, coords, preset };
     }
 
+    async loadMvsFromUrl(url: string, format: 'mvsj') {
+        if (format === 'mvsj') {
+            const data = await this.plugin.fetch({ url, type: 'string' }).run();
```
[the remark this answers is not in the document]
Comment (midlik) [reply src/apps/viewer/app.ts]: Changed.

File: src/apps/viewer/app.ts [hunk 10]
```diff
@@ -467,6 +470,30 @@ export class Viewer {
...
+            // const asset = Asset.getUrlAsset(this.plugin.managers.asset, url);
+            // const wrapper = await this.plugin.runTask(this.plugin.managers.asset.resolve(asset, 'string'));
+            // const data = wrapper.data;
+            const mvsData = MVSData.fromMVSJ(data);
```
[the remark this answers is not in the document]
Comment (midlik) [reply src/apps/viewer/app.ts:476]: Removed comments.

File: src/extensions/mvs/additions/annotation-prop.ts (hunk 1 above)
Comment (midlik) [reply src/extensions/mvs/additions/annotation-prop.ts:150]: My favorite :D 
And you can do this with it:
```
let cachedStuff = undefined;
function getStuff(){ return cachedStuff ??= actuallyComputeStuff(); }
```

File: src/extensions/mvs/helpers/param-definition.ts (hunk 5 above)
Comment (midlik) [reply src/extensions/mvs/helpers/param-definition.ts:27]: Renamed.

File: src/extensions/mvs/helpers/param-definition.ts (hunk 6 above)
Comment (midlik) [reply src/extensions/mvs/helpers/param-definition.ts:11]: Renamed.

File: src/extensions/mvs/helpers/utils.ts (hunk 7 above)
Comment (midlik) [reply src/extensions/mvs/helpers/utils.ts:121]: This is to avoid resizing `dst` multiple times when `src.length >> dst.length`

File: src/extensions/mvs/helpers/utils.ts
Comment (midlik) [reply src/extensions/mvs/helpers/utils.ts:1]: Good idea. I didn't do it because I was developing this in a separate repo, but it makes sense.

Comment (midlik) [reply src/extensions/mvs/helpers/utils.ts:1]: I moved to `mol-util/array.ts` and `mol-util/object.ts` where it made sense.

File: src/extensions/mvs/additions/annotation-color-theme.ts
Comment (midlik) [reply src/extensions/mvs/additions/annotation-color-theme.ts:1]: I couldn't come with a better name :D
`components` sounds great, I renamed it, thanks.

File: src/extensions/mvs/camera.ts (hunk 2 above)
Comment (midlik) [reply src/extensions/mvs/camera.ts:58]: Well, `this.plugin.managers.camera.focusSphere()` has a `principalAxes` option but I failed to understand what it actually does. I managed to pass some data to it, but it flips the orientation ad libitum. 
Plus, it cannot be awaited, which becomes a problem e.g. when you want to run it headless to render images. This could be solved by calling PluginCommands.Camera.Focus instead (and adding the option there), but it still doesn't solve the flipping.
To sum up, I think it's just easier to compute the snapshot and call `PluginCommands.Camera.SetSnapshot`.

File: src/extensions/mvs/helpers/utils.ts (hunk 8 above)
Comment (midlik) [reply src/extensions/mvs/helpers/utils.ts]: Moved to `mol-util/array.ts` and renamed to `arrayExtend` (following the convention in that file)

File: src/extensions/mvs/helpers/indexing.ts (hunk 3 above)
Comment (dsehnal) [reply src/extensions/mvs/helpers/indexing.ts:57]: Technically, this should be the case. But your solution does out-of-bounds access to an array and extra if check for undefined and null for each assignment.

File: src/extensions/mvs/helpers/schemas.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
+ *
+ * @author Adam Midlik <midlik@gmail.com>
+ */
+
+import { Choice } from '../../../extensions/volumes-and-segmentations/helpers';
```
[the remark this answers is not in the document]
Comment (dsehnal) [reply src/extensions/mvs/helpers/schemas.ts:7]: Yes, agree. Please move this.

[the remark this answers is not in the document]
Comment (midlik) [reply src/extensions/mvs/helpers/schemas.ts]: Agree, I moved it to mol-util/param-choice

File: src/mol-util/object.ts [hunk 12]
```diff
@@ -97,17 +98,105 @@ export function deepClone<T>(source: T): T {
...
+export function objectFromKeysAndValues<K extends keyof any, V>(keys: K[], values: V[]): Record<K, V> {
+    const obj: Partial<Record<K, V>> = {};
+    for (let i = 0; i < keys.length; i++) {
+        obj[keys[i]] = values[i];
+    }
+    return obj as Record<K, V>;
+}
+
+/** Decide if `obj` is a good old object (not array or null or other type). */
+export function isReallyObject(obj: any): boolean {
```
[the remark this answers is not in the document]
Comment (midlik) [reply src/mol-util/object.ts]: Renamed. Thank you

File: src/mol-util/object.ts [hunk 13]
```diff
@@ -97,17 +98,105 @@ export function deepClone<T>(source: T): T {
...
+    return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
+}
+
+/** Return a copy of object `obj` with sorted keys and dropped keys whose value is undefined. */
+export function sortObjectKeys<T extends {}>(obj: T): T {
```
[the remark this answers is not in the document]
Comment (midlik) [reply src/mol-util/object.ts]: My solution will always sort them alphabetically, which is what I want.
Anyway, I moved JSON related utils to mol-util/json.ts and made `sortObjectKeys` private there.

File: src/mol-util/object.ts [hunk 14]
```diff
@@ -97,17 +98,105 @@ export function deepClone<T>(source: T): T {
...
     if (!o) return;
     for (const k of Object.keys(o)) {
         f((o as any)[k], k);
     }
-}
\ No newline at end of file
+}
+
+
+/** Return `true` if object `obj` has own property with key `key` */
```
[the remark this answers is not in the document]
Comment (midlik) [reply src/mol-util/object.ts]: I'm not 100% sure if I understood your suggestion. Please check here, if it is what you meant: https://github.com/molstar/molstar/pull/976/commits/8fa19f787b3247ef86c497547ff2c375e603176b

File: src/mol-util/array.ts [hunk 15]
```diff
@@ -156,3 +158,73 @@ export function arrayMapUpsert<T>(xs: [string, T][], key: string, value: T) {
...
+export function arrayExtend<T>(dst: T[], src: ArrayLike<T>): void {
+    const offset = dst.length;
+    const nCopy = src.length;
+    dst.length += nCopy;
+    for (let i = 0; i < nCopy; i++) {
+        dst[offset + i] = src[i];
+    }
+}
+
+/** Check whether `array` is sorted, sort if not. */
```
[the remark this answers is not in the document]
Comment (midlik) [reply src/mol-util/array.ts:190]: Surprisingly, when the array is already sorted, this can be ~3x faster.
I use it when I expect the array will usually be sorted, but have to double-check (e.g. auth_seq_id).

File: src/mol-util/object.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (arose) [reply src/mol-util/object.ts]: looks good, thanks

File: src/mol-util/array.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (arose) [reply src/mol-util/array.ts:190]: cool, would you mind adding a little test to perf-tests/

**Review by dsehnal — APPROVED** [review]
Good job @midlik and @JonStargaryen working on this.

File: src/mol-util/array.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (midlik) [reply src/mol-util/array.ts:190]: @dsehnal already merged this PR, so I'll address it a future one (https://github.com/molstar/molstar/pull/991)

[the remark this answers is not in the document]
Comment (midlik) [reply src/mol-util/array.ts:190]: Addressed in this commit (https://github.com/molstar/molstar/pull/991/commits/1d200c9a8b969965b934e0b68888cd371f540996)

This is what the perf test gives:
```
sort (1000000, pre-sorted): 37.778ms
sort (1000000, pre-sorted): 38.496ms
sort (1000000, pre-sorted): 39.071ms
sortIfNeeded (1000000, pre-sorted): 2.355ms
sortIfNeeded (1000000, pre-sorted): 2.99ms
sortIfNeeded (1000000, pre-sorted): 0.992ms
sort (1000000, not sorted): 459.946ms
sort (1000000, not sorted): 422.959ms
sort (1000000, not sorted): 418.736ms
sortIfNeeded (1000000, not sorted): 437.739ms
sortIfNeeded (1000000, not sorted): 382.919ms
sortIfNeeded (1000000, not sorted): 376.24ms
```
